src/borg/archiver.py

```diff
@@ -652,126 +709,82 @@ def _process(self, *, path, parent_fd=None, name=None,
-                        try:
-                            st_target = os.stat(name, dir_fd=parent_fd, follow_symlinks=True)
-                        except OSError:
-                            special = False
-                        else:
-                            special = is_special(st_target.st_mode)
-                        if special:
-                            status = fso.process_file(path=path, parent_fd=parent_fd, name=name, st=st_target,
-                                                      cache=cache, flags=flags_special_follow)
-                        else:
-                            status = fso.process_symlink(path=path, parent_fd=parent_fd, name=name, st=st)
-            elif stat.S_ISFIFO(st.st_mode):
-                if not dry_run:
-                    if not read_special:
-                        status = fso.process_fifo(path=path, parent_fd=parent_fd, name=name, st=st)
-                    else:
-                        status = fso.process_file(path=path, parent_fd=parent_fd, name=name, st=st,
-                                                  cache=cache, flags=flags_special)
-            elif stat.S_ISCHR(st.st_mode):
-                if not dry_run:
-                    if not read_special:
-                        status = fso.process_dev(path=path, parent_fd=parent_fd, name=name, st=st, dev_type='c')
-                    else:
-                        status = fso.process_file(path=path, parent_fd=parent_fd, name=name, st=st,
-                                                  cache=cache, flags=flags_special)
-            elif stat.S_ISBLK(st.st_mode):
-                if not dry_run:
-                    if not read_special:
-                        status = fso.process_dev(path=path, parent_fd=parent_fd, name=name, st=st, dev_type='b')
-                    else:
-                        status = fso.process_file(path=path, parent_fd=parent_fd, name=name, st=st,
-                                                  cache=cache, flags=flags_special)
-            elif stat.S_ISSOCK(st.st_mode):
-                # Ignore unix sockets
-                return
-            elif stat.S_ISDOOR(st.st_mode):
-                # Ignore Solaris doors
-                return
-            elif stat.S_ISPORT(st.st_mode):
-                # Ignore Solaris event ports
-                return
-            else:
-                self.print_warning('Unknown file type: %s', path)
-                return
+                            self._rec_walk(
+                                    path=normpath, parent_fd=child_fd, name=dirent.name, fso=fso, cache=cache, matcher=matcher,
+                                    exclude_caches=exclude_caches, exclude_if_present=exclude_if_present,
+                                    keep_exclude_tags=keep_exclude_tags, skip_inodes=skip_inodes, restrict_dev=restrict_dev,
+                                    read_special=read_special, dry_run=dry_run)
+            elif right_dev:
```

Comment:
as this is never a directory, doesn't have imply it is always the right dev (or we will not even get here)?

a directory change is always at the directory mounted over the mountpoint.

Reply:
Well i just checked and linux `mount --bind` supports mounting non-directories. I'll revert it and prepare another clean patch because i believe there's an issue to be addressed there (we should probably save mount points in general, eg save only the `stat`s).

Reply:
are you sure?

it didn't work for me (i used `/rw/config/rc.local` just because it is a regular file on a different filesystem):
```
$ sudo mount --bind /rw/config/rc.local foobar
mount: mount point foobar does not exist
$ mkdir foobar
$ sudo mount --bind /rw/config/rc.local foobar
mount: mount(2) failed: /home/user/foobar: Not a directory
```

Reply:
You have to mount it onto a file. Seems kinda arbitrary: I didn't try too hard but apparently everything other than a directory can be bind-mounted onto a file. I guess the only way to know for sure would be to look at the kernel source, it's not gonna be documented anywhere.